docs/source/java/memory.rst

```diff
@@ -0,0 +1,174 @@
+.. KIND, either express or implied.  See the License for the
+.. specific language governing permissions and limitations
+.. under the License.
+
+=================
+Memory Management
+=================
+
+.. contents::
+
+The memory modules contain all the functionality that Arrow uses to manage memory (allocation and deallocation).
+This section will introduce you to the major concepts in Java’s memory management:
+
+* `BufferAllocator`_
+* `ArrowBuf`_
+* `Reference counting`_
+
+Getting Started
+===============
+
+Arrow's memory management is built around the needs of the columnar format and using off-heap memory.
+Also, it is its own independent implementation, and does not wrap the C++ implementation.
+
+Arrow offers a high level of abstraction providing several access APIs to read/write data into a direct memory.
+
+Arrow provides multiple modules: the core interfaces, and implementations of the interfaces.
+Users need the core interfaces, and exactly one of the implementations.
+
+* ``Memory Core``: Provides the interfaces used by the Arrow libraries and applications.
+* ``Memory Netty``: An implementation of the memory interfaces based on the `Netty`_ library.
+* ``Memory Unsafe``: An implementation of the memory interfaces based on the `sun.misc.Unsafe`_ library.
+
+BufferAllocator
+===============
+
+The BufferAllocator interface deals with allocating ArrowBufs for the application.
+
+The concrete implementation of the allocator is RootAllocator. Applications should generally create one RootAllocator at the
+start of the program, and use it through the BufferAllocator interface. Allocators have a memory limit. The RootAllocator
+sets the program-wide memory limit. The RootAllocator is responsible for being the master bookkeeper for memory allocations.
+
+Arrow provides a tree-based model for memory allocation. The RootAllocator is created first, then all allocators
+are created as children ``BufferAllocator.newChildAllocator`` of that allocator.
+
+One of the uses of child allocators is to set a lower temporary limit for one section of the code. Also, child
+allocators can be named; this makes it easier to tell where an ArrowBuf came from during debugging.
+
+ArrowBuf
+========
+
+ArrowBuf represents a single, contiguous allocation of `Direct Memory`_. It consists of an address and a length,
+and provides low-level interfaces for working with the contents, similar to ByteBuffer.
+
+The objects created using ``Direct Memory`` take advantage of native executions and it is decided natively by the JVM. Arrow
+offer efficient memory operations base on this Direct Memory implementation (`see section below for detailed reasons of use`).
+
+Unlike (Direct)ByteBuffer, it has reference counting built in (`see the next section`).
+
+Reference counting
+==================
+
+Direct memory involve more activities than allocate and deallocate because allocators (thru pool/cache)
+allocate buffers (ArrowBuf).
```

Comment:
Direct memory is more expensive to allocate and deallocate. That's why allocators pool or cache direct buffers. The caching is not the _reason_ why it's expensive, it's the _consequence_.

Reply:
Thank you, added on my notes. Updated